### PR TITLE
fix: bound keras zip config traversal

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary

- bound Keras `config.json` depth, admitted/pending work, direct-field work, string count, and total string characters
- build an iterative bounded projection for legacy model, layer, inbound-node, and compile metric/loss consumers
- decouple admitted structural traversal from generic string exhaustion so queued security-relevant layers remain visible
- preserve only fixed security fields after generic overflow through a separately bounded aggregate literal reserve
- keep the independently bounded unsafe-deserialization and `get_file` CVE detectors on the original config
- add constant-work direct checks for queued unsafe-deserialization nodes and direct/nested `get_file` callable shapes
- bound archive-origin slicing before whitespace normalization, regex matching, or evidence redaction
- preserve admitted scalar security fields at depth and item boundaries while omitting unadmitted values cleanly
- cap effective configurable depth at 256 so recursive legacy helpers cannot reach Python recursion limits
- reserve queued children before enqueueing, preventing nested-wide pending-work amplification
- avoid duplicate and recursive `kwargs` URL extraction across nested `get_file` calls
- stop detector loops immediately after budget exhaustion and fail closed with explicit inconclusive metadata
- retain current-main truthy `get_file` extraction semantics and their regression matrix
- remove the unused StringLookup inconclusive-reason constant reported by code quality

## Adversarial review

The review reproduced and fixed:

- wide and nested container CPU/memory amplification
- unbounded root layer, inbound-node, and compile metric/loss scans after preflight exhaustion
- benign nested `kwargs` strings being double-counted
- descendant `kwargs` subtrees being rescanned once per ancestor
- depth/item/string boundary handling erasing admitted Lambda, custom-layer, unsafe-deserialization, and `get_file` findings
- nested callable dictionaries dropping `get_file` evidence
- security-string fallback exceeding the configured aggregate character budget
- archive origins bypassing the string budget during strip/regex handling
- unknown post-overflow projection keys escaping the direct-work item bound
- synthetic `None` placeholders producing false structural evidence
- recursive projection cleanup and configured-depth recursion failures

The final read-only audit's two boundedness findings were fixed and covered by regressions.

## Validation

- complete associated Keras ZIP module: **506 passed, 73 skipped**
- adversarial traversal-budget class: **45 passed**
- affected `get_file` and unsafe-deserialization CVE classes: **64 passed**
- scoped Ruff lint and format checks
- scoped mypy
- `git diff --check`
- exact remote tree differs from current `main` in three files

Optional dependency skips require unavailable HDF5/ONNX packages. Per review policy, the full repository suite was left to CI.

Published head: `8ecc44aab1c57c701280bb61224a0b98c0eec6cd`

### CODE REVIEW

File: modelaudit/scanners/keras_zip_scanner.py [hunk 1]
```diff
@@ -1875,16 +2138,33 @@ def _check_get_file_gadget(self, model_config: Any, result: ScanResult) -> None:
...
             url_candidate_values: list[str] = []
             for key, value in node.items():
-                direct_string_values.extend(self._extract_string_literals(value))
+                value_context = self._config_child_path(context, f".{redact_evidence_string(str(key), max_chars=64)}")
+                direct_string_values.extend(
+                    self._extract_string_literals(
+                        value,
+                        result=result,
+                        state=state,
+                        context=value_context,
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Confirmed and reproduced. The local validated fix separates `_iter_dict_nodes()` accounting from string-extraction accounting, adds an explicit halted state for global item/string exhaustion, and preserves current-main effective `get_file` argument semantics. Regression: `test_scalar_siblings_within_budget_do_not_hide_nested_get_file_cve` covers both CVE-2025-8747 and CVE-2025-12060. Local integrated head `fe515d68`; 50 focused Keras tests plus Ruff/format/mypy pass. Shell GitHub auth cannot publish the commit, so @codex address that feedback using this implementation and regression.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 2]
```diff
@@ -1875,16 +2138,33 @@ def _check_get_file_gadget(self, model_config: Any, result: ScanResult) -> None:
...
+                    )
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Confirmed and reproduced with an instrumented shallow dict. The local validated fix stops the per-node key/value loop once the detector's independent literal-work state is halted, while traversal has its own bounded state. Regression: `test_shallow_dict_iteration_stops_after_detector_budget` covers both get-file detectors and fails on reviewed head `2447f8e2`. Local integrated head `fe515d68`; 50 focused Keras tests plus Ruff/format/mypy pass. Shell GitHub auth cannot publish the commit, so @codex address that feedback using this implementation and regression.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 3]
```diff
@@ -1875,16 +2138,33 @@ def _check_get_file_gadget(self, model_config: Any, result: ScanResult) -> None:
...
         Detected when a single config object references get_file and includes URL arguments.
         """
-        for context, node in self._iter_dict_nodes(model_config):
+        state = self._new_config_traversal_state()
+        for context, node in self._iter_dict_nodes(model_config, result, state=state):
             if self._is_primarily_documentation(context, node):
                 continue
             direct_string_values: list[str] = []
             url_candidate_values: list[str] = []
             for key, value in node.items():
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Confirmed; this is the same reproduced CPU-bound failure covered by local integrated head `fe515d68`. The fix adds `_ConfigTraversalState.halted`, sets it on item/string exhaustion, uses independent traversal and literal states, and checks `literal_state.halted` before each yielded-node entry so both get-file detectors break immediately. `test_shallow_dict_iteration_stops_after_detector_budget` instruments both `items()` and `values()` and fails if either detector exceeds bounded work. 50 focused Keras tests plus Ruff/format/mypy pass. @codex address that feedback with that implementation; keep the thread open until the remote head changes.

File: modelaudit/scanners/keras_zip_scanner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Fixed in 8d6fa15. The two get_file detectors now use independent traversal and literal-accounting states, so parent scalar extraction no longer consumes the iterator's queue budget. The nested CVE regression covers both CVE-2025-8747 and CVE-2025-12060.

File: modelaudit/scanners/keras_zip_scanner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Fixed in 8d6fa15. Item/string exhaustion sets an explicit halted state, and each yielded-dict loop checks it before processing another entry. The instrumented shallow-dict regression covers both get_file detectors and enforces bounded iteration.

File: modelaudit/scanners/keras_zip_scanner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Fixed in 8d6fa15. Both detector loops now break as soon as their independent literal state is halted; the underlying iterator also stops globally on item exhaustion. Focused and full Keras ZIP tests pass.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 4]
```diff
@@ -897,6 +939,188 @@ def _mark_inconclusive_scan_result(result: ScanResult, reason: str) -> None:
...
+            if not self._reserve_config_traversal_item(state, result, context=context, depth=depth):
+                continue
+
+            if isinstance(node, str):
+                self._record_config_string_literal(node, state, result, context=context)
+                continue
+
+            if isinstance(node, dict):
+                remaining_items = max(state.max_items - state.items_seen, 0)
+                for queued_items, (key, value) in enumerate(node.items()):
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Confirmed and fixed locally in `7678499d` on top of current `main`. The implementation adds monotonic `items_reserved` accounting, reserves the root and every child before enqueueing, increments `items_seen` only when dequeued, and applies the same invariant to `_validate_config_traversal_budget`, nested module traversal, unsafe-deserialization traversal, `_extract_string_literals`, and `_iter_dict_nodes`. `test_nested_wide_containers_reserve_queued_items` exercises all five paths; `test_exact_item_budget_does_not_fail_closed` protects the benign boundary. Validation: 15 focused budget tests and all 539 associated Keras ZIP tests passed; repository Ruff format/check and mypy (464 files) passed. Shell GitHub credentials cannot push, so @codex address that feedback using this implementation and tests. Keep this thread open until the remote head changes.

[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Fixed in remote commit `fe046f2191da0f8675d8cc73f36c21ac3d958eaf`. Queue entries are now reserved before enqueue, pending work counts against the global item budget, already-admitted nodes remain inspectable after overflow, and direct-field work has its own global bound. Added adversarial wide/nested regressions plus last-admitted malicious and benign boundary cases. Validation: 20 traversal-budget tests passed; full associated Keras ZIP module 471 passed, 73 optional-dependency skips; Ruff, format, mypy, and diff checks clean.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 5]
```diff
@@ -897,6 +939,188 @@ def _mark_inconclusive_scan_result(result: ScanResult, reason: str) -> None:
...
+                    child_context = self._config_child_path(context, f"[{queued_items}]")
+                    if queued_items >= remaining_items:
+                        self._reserve_config_traversal_item(
+                            state,
+                            result,
+                            context=child_context,
+                            depth=depth + 1,
+                        )
+                        break
+                    pending.append((value, depth + 1, child_context))
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Fixed in remote commit `fe046f2191da0f8675d8cc73f36c21ac3d958eaf`. Queue entries are now reserved before enqueue, pending work counts against the global item budget, already-admitted nodes remain inspectable after overflow, and direct-field work has its own global bound. Added adversarial wide/nested regressions plus last-admitted malicious and benign boundary cases. Validation: 20 traversal-budget tests passed; full associated Keras ZIP module 471 passed, 73 optional-dependency skips; Ruff, format, mypy, and diff checks clean.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 6]
```diff
@@ -239,7 +242,14 @@ def _archive_format_may_extract_tar(value: Any) -> bool:
 _HDF5_USERBLOCK_MAX_CONCATENATED_ZIP_SEGMENTS = 16
 _ZIP_END_OF_CENTRAL_DIRECTORY_SIGNATURE = b"PK\x05\x06"
 _ZIP_END_OF_CENTRAL_DIRECTORY_FIXED_BYTES = 22
-_MAX_STRING_LITERAL_EXTRACTION_DEPTH = 100
+_MAX_CONFIG_TRAVERSAL_PATH_CHARS = 512
+_CONFIG_TRAVERSAL_DEPTH_EXCEEDED_REASON = "keras_zip_config_traversal_depth_exceeded"
+_CONFIG_TRAVERSAL_ITEM_LIMIT_EXCEEDED_REASON = "keras_zip_config_traversal_item_limit_exceeded"
+_CONFIG_STRING_LITERAL_LIMIT_EXCEEDED_REASON = "keras_zip_config_string_literal_limit_exceeded"
+_CONFIG_STRING_CHAR_LIMIT_EXCEEDED_REASON = "keras_zip_config_string_char_limit_exceeded"
+_KERAS_STRINGLOOKUP_EXTERNAL_VOCABULARY_INCONCLUSIVE_REASON = (
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Confirmed and fixed locally. The unused constant is removed, and the review also found a broader false-negative/DoS gap: budget validation was bounded, but downstream layer, compile-config, and evidence analysis still received the original unbounded object. Local commit `42b9a2c9` adds an order-preserving bounded projection for every structured analysis path and regressions proving wide configs are truncated while an admitted malicious `get_file` prefix is still detected. Integrated with current `main` in `c1a426bb`; 556 associated Keras ZIP tests pass, plus repository Ruff, format, mypy, and diff checks. Shell GitHub auth cannot push, so @codex address that feedback using these commits and keep this thread open until the remote head changes.

**Comment by mldangelo-oai** [comment]
Critical review found and fixed an additional bounded-traversal bypass beyond the existing review threads: `_validate_config_traversal_budget()` reported overflow, but `_scan_model_config()` and compile/evidence helpers still traversed the original unbounded config. Local commit `42b9a2c9` routes all downstream structured analysis through an order-preserving bounded projection, preserves admitted malicious prefixes, removes the unresolved unused constant, and adds focused regressions. Current-main integrated head: `c1a426bb`.

Validation:
- `556 passed` in `tests/scanners/test_keras_zip_scanner.py`
- repository Ruff check clean
- repository Ruff format check clean
- repository mypy clean (464 files)
- `git diff --check` clean
- Published Python CI failure is unrelated: `TestWeightDistributionScanner::test_pytorch_model_scan`, fixed on current main by #1593

Shell GitHub credentials reject pushes in this environment. @codex apply/push local commits `42b9a2c9` and `c1a426bb` to `mdangelo/codex/fix-keras-config-traversal-budget`, then resolve the remaining code-quality thread after the remote head updates.

**Comment by mldangelo-oai** [comment]
Completed the final critical review and published the exact validated tree at `c0b536989aec902b878ad3c4dd4f147d7c8c1596`.

In addition to removing the unused constant from the live code-quality thread, this closes the remaining traversal gaps found during adversarial QA: downstream layer/inbound/compile consumers now receive an iterative bounded projection; independently bounded CVE detectors retain original-config boundary visibility; nested kwargs are not double-counted or recursively rescanned; admitted scalar security fields survive depth/item exhaustion; omitted fields do not create synthetic `None` evidence; and configurable depth is capped at the recursion-safe 256 ceiling.

Final associated validation: 492 passed, 73 optional-dependency skips; 31 focused adversarial budget tests; Ruff, formatting, mypy, and diff checks clean. All review threads are resolved, and the remote diff is exactly the intended three files.

**Comment by mldangelo-oai** [comment]
Published final validated head `793040ae`.

Late review findings addressed:
- admitted nested `get_file` gadgets remain detectable after one or multiple literal overflows
- queued unsafe-deserialization nodes remain detectable after overflow
- queued Lambda/custom-layer findings survive generic string exhaustion through a separately bounded security reserve
- nested callable dictionaries preserve exact `get_file` references without matching helper near-matches
- archive origins are capped before strip/regex/redaction
- post-overflow security strings and unknown projection keys remain aggregate-budget bounded

Validation on the exact published tree:
- Keras ZIP scanner module: **506 passed, 73 skipped**
- traversal-budget regressions: **45 passed**
- affected CVE classes: **64 passed**
- scoped Ruff, format, mypy, and `git diff --check`: clean
